File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N = int(input())
graph = [[] for _ in range(N + 1)]

for _ in range(N - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

parents = [0] * (N + 1)

def dfs(node):
    for child in graph[node]:
        if parents[child] == 0:
            parents[child] = node
            dfs(child)
dfs(1)

print('\n'.join(map(str, parents[2:])))

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.js
const input = require("fs")
  .readFileSync(0)
  .toString()
  .trim()
  .split("\n")
  .map((v) => v.trim());

const N = parseInt(input[0]);
const graph = {};

for (let i = 1; i <= N; i++) {
  graph[i] = [];
}

for (let i = 1; i < N; i++) {
  const [a, b] = input[i].split(" ").map(Number);
  graph[a].push(b);
  graph[b].push(a);
}

const parents = new Array(N + 1).fill(0);

const dfs = (node) => {
  for (const child of graph[node]) {
    if (parents[child] === 0 && child !== 1) {
      parents[child] = node;
      dfs(child);
    }
  }
};

dfs(1);

console.log(parents.slice(2).join("\n"));
